app/utils/graphql.ts

```diff
@@ -27,40 +29,74 @@ const DEFAULT_SWR_OPTIONS: SWRConfiguration = {
 /**
  * Fetcher which interacts with our GraphQL API
  */
-const fetcher = <T>([query, variables, token, hasuraRoleName]: [
+const fetcher = <T>([query, variables, token, hasuraRoleName, schema]: [
   RequestDocument,
   Variables,
   string,
-  string
+  string,
+  ZodSchema<T>
 ]): Promise<T> =>
-  request({
+  request<T>({
     url: ENDPOINT,
     document: query,
     variables,
     requestHeaders: {
       Authorization: `Bearer ${token}`,
       "x-hasura-role": hasuraRoleName,
     },
+  }).then((data) => {
+    return schema.parseAsync(data);
   });
 
+/**
+ * Hook which wraps our data schema in an outer object that
+ * matches what Hasura will return. e.g., it shapes a Crash
+ * object schema into a shape like
+ * { crashes: <Crash[]> }
+ */
+const useApiResponseSchema = <T extends Record<string, unknown>>(
+  schema: ZodSchema<T>,
+  typename: string
+): ZodSchema<{ [key in typeof typename]: T[] }> =>
+  useMemo(() => {
+    return z.object({ [typename]: z.array(schema) });
+  }, [schema, typename]);
+
+interface UseQueryResponse<T> {
+  data?: T[];
+  error: any;
+  isLoading: boolean;
+  isError: boolean;
+  refetch: () => Promise<any>;
+  isValidating: boolean;
+}
 /**
  * Hook which wraps `useSWR` and provides auth
+ *
+ * This hook is limited to querying just one type per query as
+ * will be defined by the typename
```

Comment:
what i'm trying to say is that this `useQuery()` hook only supports fetching one type of object (crashes, units, etc) at the root of the query. 

so this is fine:

```gql
 query {
    crashes {
      id
    }
  }
}
```

but this is not:

```gql
query {
    crashes {
      id
    }
      units {
        id
      }
  }
```

we could extend this to support multiple root types but i wasn't up for figuring out how we would infer types from this kind of shape.